### DISCUSSION TITLE
[General] Paymaster for sending stable coins (USDC,USDT,DAI) without ETH

### DISCUSSION THREAD

**Discussion by moltam89** [zkSync-Community-Hub/zksync-developers#46] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Hi guys,

I'd like to extend the functionality of the wallet I'm working on, allowing users to send stable coins (USDC,USDT,DAI) without ETH.

Do I need to build my own paymaster for this?
Are there any "official" paymasters I can use to pay the transaction fee with ERC20 tokens?

Thanks,
Tamas

### What is the specific issue or error you're encountering?

-

### Can you share the error messages or logs you're receiving, if any?

-

### Have you made any recent changes to the contract before encountering this issue?

-

### Are there any external libraries or contracts that your contract interacts with?

-

### Can you provide the relevant portions of your contract code where the issue is occurring?

-

### Have you tried to isolate the problem, and if so, what were the results?

-

### What steps have you already taken to try to resolve the issue?

-

**Comment by uF4No** [comment, marked as answer]
Yes, you'd need your own paymaster. The paymaster will use ETH to pay the transaction fees behind the scenes, but you can make your users pay only using USDC or any other token. You'd need to create a paymaster contract that implements the [IPaymaster interface](https://era.zksync.io/docs/reference/concepts/aa.html#ipaymaster-interface) and, as you want to use ERC20 tokens, use the [ApprovalBased paymaster flow](https://era.zksync.io/docs/reference/concepts/aa.html#built-in-paymaster-flows).

Take a look at our [Quickstart](https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html) that uses the testnet paymaster that allows users to pay fees in different ERC20 tokens. Note that the testnet paymaster is for demonstration purpose only and that the exchange rate it uses is fixed 1 wei for 1 unit of a token.

Once you're familiar with that, you can also follow the [custom paymaster tutorial](https://era.zksync.io/docs/dev/tutorials/custom-paymaster-tutorial.html) and the [USDC API3 paymaster tutorial](https://era.zksync.io/docs/dev/tutorials/api3-usd-paymaster-tutorial.html)

**Reply by moltam89** [reply]
Thanks @uF4No !

I was hoping for ready to use paymaster. :)

Does this mean that each wallet which would like to implement gasless transactions will have their own paymaster?

**Reply by uF4No** [reply]
Yes because, even if the transactions are gasless for the users, the paymaster is actually paying the gas fees in ETH. So someone has to make sure that the paymaster has enough ETH to pay those gas fees for the users. 

If there was a "common" paymaster for multiple wallets, someone (or multiple people) have to make sure that the paymaster balance is monitored and topped up when needed. If the paymaster runs out of ETH, it just stops working.

**Reply by moltam89** [reply]
Thank you @uF4No !

**Reply by moltam89** [reply]
Hi @uF4No 
Follow up question:

Is there maybe a Live example for an erc20 paymaster?
It doesn't seem straightforward how to get the prices (I got a tip from Ramon that [RedStone oracles](https://github.com/redstone-finance/zkSync-tutorial-stable-price-marketplace/blob/main/README.md) might be help), it would be nice to look at a working example.

Thanks,
Tamas

**Reply by uF4No** [reply]
In the [USDC API3 paymaster tutorial](https://era.zksync.io/docs/dev/tutorials/api3-usd-paymaster-tutorial.html) you can see how it uses API3 to fetch the ETH/USD prices. We don't have an example with Redstone (yet!) but it should be similar.

**Reply by moltam89** [reply]
Hi @uF4No ,

I created a [USDC paymaster](https://github.com/moltam89/scaffold-eth-2/blob/dbc949fa81f0f606ecb0011823a926dd78b8b136/packages/hardhat/contracts/MyPaymaster.sol) in zkSyncTestnet, however I wasn't able to calculate the required USDC amount properly because of the gas refunds.

For example in this [transaction](https://goerli.explorer.zksync.io/tx/0x3dfb814931ca8f97de66345e555506e7279f3f0faec9f342acf80b58c9bdee1b), the initial fee is 0.001103566 ETH (*1888 = 2.083$) and therefore 2.083 USDC is transferred to the paymaster, but later most of the tx fee is refunded.

I was following the mentioned [USDC API3 paymaster tutorial](https://era.zksync.io/docs/dev/tutorials/api3-usd-paymaster-tutorial.html), here is the [related code](https://github.com/moltam89/scaffold-eth-2/blob/dbc949fa81f0f606ecb0011823a926dd78b8b136/packages/hardhat/contracts/MyPaymaster.sol#L79-L92) part from my test paymaster.

Is there a better way to determine the required token amount?

Best,
Tamas

**Comment by martinerinv** [comment]
To enable users to send stable coins (such as USDC, USDT, DAI) without requiring ETH for transaction fees, you can integrate a paymaster solution into your wallet application. Paymasters facilitate the payment of transaction fees in tokens other than ETH.

You have a few options:

Build your own paymaster: Develop a custom paymaster contract that accepts stable coins as payment for transaction fees and integrates it into your wallet application.
Use existing paymasters: Some projects and platforms provide paymaster contracts that you can utilize in your application. Check if there are any official or trusted paymasters available that support ERC20 tokens for paying transaction fees.
Ensure that whichever option you choose complies with security best practices and aligns with your project's requirements and architecture. Additionally, consider the gas efficiency and reliability of the chosen paymaster solution.

**Comment by evmeiwear** [comment]
To enable users to send stable coins (USDC, USDT, DAI) without ETH for transaction fees, you can utilize paymasters. You may either build your own paymaster or explore existing "official" paymasters that support paying transaction fees with ERC20 tokens.

**Comment by mcleodjv** [comment]
Certainly! To enable users to send stablecoins (USDC, USDT, DAI) without requiring ETH for transaction fees, you can implement a custom paymaster contract. This paymaster contract would accept ERC20 tokens as payment for transaction fees and handle the execution of transactions on behalf of users.

As for "official" paymasters, there isn't a standard or official paymaster provided by zkSync. You would need to build your own paymaster contract tailored to your specific requirements.

Here's a high-level

**Comment by lashawnheisler** [comment]
To enable users to send stable coins without ETH for transaction fees, you can utilize a paymaster contract. You can either build your own paymaster contract or use existing ones provided by the community. Check for available paymaster contracts that support paying transaction fees with ERC20 tokens.

**Comment by hyal1ngrumblu** [comment]
The user is seeking guidance on extending the functionality of their wallet to allow users to send stable coins without ETH on the Mainnet. They are asking if they need to build their own paymaster for this, or if there are any "official" paymasters they can use to pay the transaction fee with ERC20 tokens. The user has not provided any specific issue or error messages. They have not made any recent changes to their contract and it is unclear if they are interacting with any external libraries or contracts. The user has not provided any relevant portions of their contract code. The user is requesting assistance with resolving this issue.
